experimental/tzif/src/core/mod.rs

```diff
@@ -0,0 +1,151 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+/// Generic trait extensions to the core parsing functionality.
+pub mod ext;
+
+/// A result containing a parsed item and a source for the remaining input.
+pub type ParseResult<Item, Source> = eyre::Result<Parsed<Item, Source>>;
+
+/// A parse error.
+#[derive(thiserror::Error, Debug)]
+pub enum ParseError {
+    /// The error value that is left behind when a populated error is
+    /// moved to a subsequent parse result.
+    #[error("empty error")]
+    Empty,
+    /// The source has reached the end of its input.
+    #[error("end of parse input")]
+    EndOfInput,
+    /// There is no source from which to draw more input.
+    #[error("no source")]
+    NoSource,
+}
+
+/// A parsed value and a source that contains the remaining parse input.
+#[derive(Debug, Clone)]
+pub struct Parsed<Value, Source: Clone>(Value, Source);
+
+impl<Value, Source: Clone> Parsed<Value, Source> {
+    /// Creates a new [`Parsed`] object from a value and a source.
+    pub fn new(value: Value, source: Source) -> Self {
+        Self(value, source)
+    }
+
+    /// Turns this [`Parsed`] object into its value, dropping the source.
+    pub fn into_value(self) -> Value {
+        self.0
+    }
+
+    /// Returns a reference of this [`Parsed`] object's value.
+    pub fn value(&self) -> &Value {
+        &self.0
+    }
+
+    /// Returns a copy of this [`Parsed`] object's source.
+    pub fn source(&mut self) -> Source {
+        self.1.clone()
+    }
+
+    /// Splits this [`Parsed`] object into a tuple of its value and source.
+    pub fn split(self) -> (Value, Source) {
+        (self.0, self.1)
+    }
+}
+
+/// A trait describing a type that is able to provide parse input.
+/// `Item` is the type of item returned by [`ParseInput::next()`].
+/// `Collection` is the type of collection of multiple items returned by [`ParseInput::take()`]
+/// `Source` is a clonable source of parse input.
+pub trait ParseInput<Item, Collection, Source = Self>
+where
+    Source: Clone + ParseInput<Item, Collection, Source>,
+    Self: Sized,
```

Comment:
Suggestion: Make the trait specific to what you need. Everywhere in this crate, you use the same type parameters:

```
ParseInput<u8, Vec<u8>, Source>
```

If this is all you need, then I would just make the ParseInput trait always assume those types, removing the type parameters.